File: Семинар наставника 2/DZ/tex/src/QuickSortSeq.cpp
void swap(int* a, int* b)
{
  if (*a == *b) return;
  *a = *a + *b;
  *b = *a - *b;
  *a = *a - *b;
}

int partitionLomuto(int* arr, int low, int high)
{
  int pivot = arr[high];
  int i = (low - 1);

  for (int j = low; j < high; j++)
    if (arr[j] <= pivot)
    {
      i++;
      swap(&arr[i], &arr[j]);
    }

  swap(&arr[i + 1], &arr[high]);
  return (i + 1);
}

int partitionHoar(int* arr, int low, int high)
{
  size_t i = low;
  size_t j = high;
  int pivot = arr[(i + j) / 2];
  while (1)
  {
    while (arr[i] < pivot)
      ++i;
    while (arr[j] > pivot)
      --j;

    if (i >= j)
      break;

    swap(&arr[i++], &arr[j--]);
  }
  return j;
}

void quickSortSeq(int* arr, int low, int high)
{
  if (low < high)
  {
    int pi = partitionHoar(arr, low, high);
    // Рекурсивно сортируем элементы, находящиеся до пивота
    quickSortSeq(arr, low, pi);
    // Рекурсивно сортируем элементы, находящиеся после пивота
    quickSortSeq(arr, pi + 1, high);
  }
}

File: Семинар наставника 2/DZ/tex/src/QuickSortSeq.py
import numpy as np
from numba import njit
import time

def quickSortSeq(arr, low, high):
    if (low < high):
        # partitionHoar
        i = low
        j = high
        pivot = arr[(i + j) // 2]
        while (1):
            while (arr[i] < pivot):
                i = i + 1
            while (arr[j] > pivot):
                j = j - 1

            if (i >= j):
                break

            arr[i], arr[j] = arr[j], arr[i]
            i = i + 1
            j = j - 1
        pi = j
        # recurcive call
        quickSortSeq(arr, low, pi)
        quickSortSeq(arr, pi + 1, high)

@njit
def quickSortSeqNjit(arr, low, high):
    if (low < high):
        # partitionHoar
        i = low
        j = high
        pivot = arr[(i + j) // 2]
        while (1):
            while (arr[i] < pivot):
                i = i + 1
            while (arr[j] > pivot):
                j = j - 1

            if (i >= j):
                break

            arr[i], arr[j] = arr[j], arr[i]
            i = i + 1
            j = j - 1
        pi = j
        # recurcive call
        quickSortSeqNjit(arr, low, pi)
        quickSortSeqNjit(arr, pi + 1, high)
        
@njit
def quickSortIterative(arr):
    stack = []
    stack.append((0, len(arr) - 1))

    while stack:
        low, high = stack.pop()

        if low < high:
            # partitionHoar
            i = low
            j = high
            pivot = arr[(i + j) // 2]
            while (1):
                while (arr[i] < pivot):
                    i = i + 1
                while (arr[j] > pivot):
                    j = j - 1
    
                if (i >= j):
                    break
    
                arr[i], arr[j] = arr[j], arr[i]
                i = i + 1
                j = j - 1
            pivot = j
            # using stack
            stack.append((low, pivot))
            stack.append((pivot + 1, high))